Feature #224: Activate tokens automatically

fernandomg opened the pull request:
This PR closes #224, by:

* Adding:
  - a `blacklistedTokens` list to the safe's store
  - a `BLACKLISTED_TOKENS` key to the persistence layer (`immortal`)
  - all the required actions, selectors, reducers required to properly update _blacklisted tokens_
  - a specific action `activeTokensByBalance` which implements the rules to:
    - persist the list of `activeTokens`
    - add all the tokens that have `balance > 0`
    - avoid those tokens that were _blakclisted_

* Editing:
  - how/when `Balances` and `TokenList` components loads (fetches) tokens
    - now `Balances` fetches tokens from backend every time it is loaded
    - `TokenList` no longer fetches tokens but gets those provided by the `Balances` component

![blacklisting-tokens](https://user-images.githubusercontent.com/3315606/69595215-b8276500-0fdd-11ea-8f1b-6d311ce3d1eb.gif)


**Note:** about the new endpoint described in [#116](https://github.com/gnosis/safe-relay-service/issues/116), despite it's listed in the swagger documentation, and it seems to be implemented in the `safe-relay-service` repo's `master` branch. I wasn't able to access it. I'm getting a `404` (https://safe-relay.staging.gnosisdev.com/api/v1/safes/0x5bF46392B1f0C3C74Df9EB70DC8a6162C71EeC28/balances/)

gnosis-info commented:
Travis automatic deployment:
 https://pr300--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr300--safereactstorybook.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr300--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr300--safereactstorybook.review.gnosisdev.com

fernandomg commented:
@lukasschor, I have added a modification on how it calculates the "to be active" tokens based on the balance.

The version you tested took the whole list of available tokens returned by the server [`/v1/tokens/`](https://safe-relay.gnosis.pm/#operations-v1-v1_tokens_list) and checked its balances using web3.

The last change I added uses the list of tokens with their balances returned by [`/v1/safes/{address}/balances/`](https://safe-relay.gnosis.pm/#operations-v1-v1_safes_balances_list)

This simplifies a lot the process, does fewer requests and uses the endpoint stated in the issue.

mikheevm commented on src/logic/tokens/store/actions/activeTokensByBalance.js:
```diff
@@ -0,0 +1,46 @@
+import fetchTokenBalanceList from '~/logic/tokens/api/fetchTokenBalanceList'
+import updateSafe from '~/routes/safe/store/actions/updateSafe'
+
+const activeTokensByBalance = (safeAddress: string) => async (dispatch: ReduxDispatch<GlobalState>, getState: GetState) => {
```
since this is a function its name should include a verb, my suggestion is `activateTokensByBalance`

fernandomg replied:
indeed, thanks!

I changed it to the suggested name. Also fixed a couple of linting errors about _line too long_ and _no return value_ for the same action.

mikheevm replied:
Thank you!

mikheevm commented on src/logic/tokens/utils/tokensStorage.js:
```diff
@@ -24,6 +25,20 @@ export const getActiveTokens = async (): Promise<Object<string, TokenProps>> =>
   return data || {}
 }
 
+export const saveBlacklistedTokens = async (tokens: Map<string, Token>) => {
```
why do we store them in a separate key? Aren't they stored individually for each safe in safe's record?

mikheevm replied:
So we store active tokens for all safes because their data should always be there for better UX. But I think for blacklisted ones it doesn't make much sense

mikheevm replied:
@fernandomg what about this?

fernandomg replied:
@mikheevm, sorry for the late response. You can see the changes in https://github.com/gnosis/safe-react/pull/300/commits/b0791301e464ad302707282342f0f59f1ef636cd

By deleting the action to persist _blakclisted tokens_ into localStorage it finished working. It took me a while to get around it and understand what I was doing wrong.

Let me know if there's anything else to fix or change.

Thank you!

mikheevm replied:
Thanks :)

gnosis-info commented:
Travis automatic deployment:
 https://pr300--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr300--safereactstorybook.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr300--safereact.review.gnosisdev.com
